### PR TITLE
An extra sandbox container created after job pod succeed

### LINKED ISSUE #75441: An extra sandbox container created after job pod succeed
/sig node
/area kubelet

@dchen1107
@derekwaynecarr 

**What happened**:

I created a cronjob in a kubernetes cluster(version 1.13.1),  the configuration is shown  below:
```
apiVersion: batch/v1beta1
kind: CronJob
metadata:
  name: hello
spec:
  schedule: "*/1 * * * *"
  jobTemplate:
    spec:
      template:
        spec:
          containers:
          - name: hello
            image: busybox
            args:
            - /bin/sh
            - -c
            - date; echo Hello from the Kubernetes cluster
          restartPolicy: OnFailure

```
Every  minute kubernetes will create a job pod. But  when a job pod completes, sometimes I can find an extra sandbox container which runs the pause program. Here is an example,  job pod hello-1552914120-qvkpr created another sandbox container after the main container busybox completed and it's initial sandbox container was killed.

```
[root@10-9-187-57 ~]# docker ps -a
CONTAINER ID        IMAGE                                                                                       COMMAND                  CREATED             STATUS                       PORTS               NAMES
d69f9a0aac47        uhub.service.ucloud.cn/google_containers/pause-amd64:3.0                                    "/pause"                 4 seconds ago       Exited (137) 1 second ago                        k8s_POD_hello-1552914120-qvkpr_default_0920a1ff-497e-11e9-8aae-525400fa7819_1
cc527083a48f        docker.io/busybox@sha256:061ca9704a714ee3e8b80523ec720c64f6209ad3f97c0ff7cb9ec7d19f15149f   "/bin/sh -c 'date;..."   5 seconds ago       Exited (0) 4 seconds ago                         k8s_hello_hello-1552914120-qvkpr_default_0920a1ff-497e-11e9-8aae-525400fa7819_0
0abe74d2aa0f        uhub.service.ucloud.cn/google_containers/pause-amd64:3.0                                    "/pause"                 19 seconds ago      Exited (0) 4 seconds ago                         k8s_POD_hello-1552914120-qvkpr_default_0920a1ff-497e-11e9-8aae-525400fa7819_0
```
What's worse is that after the second sandbox container killed by kubelet, the CNI plugin cannot get the correct NETNS Path(/proc/$pausepid/net/ns) when it tries to tear down the network which is created for the extra second sandbox container. That will definitely result in IP leakage when the relationships between pod ip and pod name, namespace as well as sandbox container id,  are not stored in somewhere else.

**What you expected to happen**:

When a regular job pod completes and becomes Succeeded, no more extra sandbox containers will be created.

**How to reproduce it (as minimally and precisely as possible)**:

Just deploy an cronjob like this:

```
apiVersion: batch/v1beta1
kind: CronJob
metadata:
  name: hello
spec:
  schedule: "*/1 * * * *"
  jobTemplate:
    spec:
      template:
        spec:
          containers:
          - name: hello
            image: busybox
            args:
            - /bin/sh
            - -c
            - date; echo Hello from the Kubernetes cluster
          restartPolicy: OnFailure
```
And perform `docker ps -a` command on a node which is sheduled to run a single job.
You are likely to find the extra sandbox of the job pod.

**Anything else we need to know?**:
After some investigation, I suppose that something is wrong when SyncPod function starts.
https://github.com/kubernetes/kubernetes/blob/v1.13.1/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L585

It ignores the fact that the pod's status is completed,  becomes SUCCEEDED, and still creates an extra sandbox containers after the first sandbox is killed and related PLEG is activated.

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.1", GitCommit:"eec55b9ba98609a46fee712359c7b5b365bdd920", GitTreeState:"clean", BuildDate:"2018-12-13T10:39:04Z", GoVersion:"go1.11.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.1", GitCommit:"eec55b9ba98609a46fee712359c7b5b365bdd920", GitTreeState:"clean", BuildDate:"2018-12-13T10:31:33Z", GoVersion:"go1.11.2", Compiler:"gc", Platform:"linux/amd64"}

```

- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
```
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
```
- Kernel (e.g. `uname -a`):
```
Linux 10-9-29-155 4.4.169-1.el7.elrepo.x86_64 #1 SMP Fri Dec 21 09:54:06 EST 2018 x86_64 x86_64 x86_64 GNU/Linux
```

- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by liucimin**
**What type of PR is this?**

 /kind bug

**What this PR does / why we need it**:

When a job pod finish quickly.
Such as the container's args is "date; echo Hello from the Kubernetes cluster"
the container and sandbox will exited quickly,
then sometimes if the RECONCILE event slower than the PLEG ContainerDied event kubelet will start a new sandbox.
It because the kubelet worker use the old pod status from the podmanager 
So it will create two sandbox when a job finish.
The first sandbox will teardown when in the PLEG ContainerDied Even syncpod.
The second sandbox will teardown by housekeeping.
But the second sandbox dont have network namespace.So the cni can not release some network 
resource.
So we should try to not create two sandbox in one job,

Check if the pod's container all succeeded,the kubelet will not create a new sandbox.

**Which issue(s) this PR fixes**:

Fixes #75441 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
NONE

```release-note
When check sandbox changed,filter the succeeded pod.
```

### CODE REVIEW

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go [hunk 1]
```diff
@@ -417,25 +417,39 @@ func (m *kubeGenericRuntimeManager) podSandboxChanged(pod *v1.Pod, podStatus *ku
 		}
 	}
 
-	// Needs to create a new sandbox when readySandboxCount > 1 or the ready sandbox is not the latest one.
+	//check the pod is succeeded
+	succeededContainerCount := 0
+	for _, container := range pod.Spec.Containers {
+		for _, status := range podStatus.ContainerStatuses {
```
Comment (tallclair) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go]: Why not just iterate over the ContainerStatuses?

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go [hunk 2]
```diff
@@ -417,25 +417,39 @@ func (m *kubeGenericRuntimeManager) podSandboxChanged(pod *v1.Pod, podStatus *ku
...
+
+	// Needs to create a new sandbox when readySandboxCount > 1.
 	sandboxStatus := podStatus.SandboxStatuses[0]
 	if readySandboxCount > 1 {
 		klog.V(2).Infof("More than 1 sandboxes for pod %q are ready. Need to reconcile them", format.Pod(pod))
 		return true, sandboxStatus.Metadata.Attempt + 1, sandboxStatus.Id
 	}
-	if sandboxStatus.State != runtimeapi.PodSandboxState_SANDBOX_READY {
+
+	//Needs to create a new when not all container succeeded and the ready sandbox is not the latest one
```
Comment (tallclair) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go]: Space after //
create a new what?
```suggestion
	// Needs to create a new ??? when not all container succeeded and the ready sandbox is not the latest one
```

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go [hunk 3]
```diff
@@ -417,25 +417,39 @@ func (m *kubeGenericRuntimeManager) podSandboxChanged(pod *v1.Pod, podStatus *ku
...
+	if len(pod.Spec.Containers) != succeededContainerCount && sandboxStatus.State != runtimeapi.PodSandboxState_SANDBOX_READY {
 		klog.V(2).Infof("No ready sandbox for pod %q can be found. Need to start a new one", format.Pod(pod))
 		return true, sandboxStatus.Metadata.Attempt + 1, sandboxStatus.Id
 	}
 
-	// Needs to create a new sandbox when network namespace changed.
-	if sandboxStatus.GetLinux().GetNamespaces().GetOptions().GetNetwork() != networkNamespaceForPod(pod) {
+	// Needs to create a new sandbox when not all container succeeded and network namespace changed.
```
Comment (tallclair) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go]: nit: I would rephrase this as
```suggestion
	// Needs to create a new sandbox when network namespace changed, unless all containers are succeeded.
```

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go [hunk 4]
```diff
@@ -417,25 +417,39 @@ func (m *kubeGenericRuntimeManager) podSandboxChanged(pod *v1.Pod, podStatus *ku
...
+	// Needs to create a new sandbox when readySandboxCount > 1.
 	sandboxStatus := podStatus.SandboxStatuses[0]
 	if readySandboxCount > 1 {
 		klog.V(2).Infof("More than 1 sandboxes for pod %q are ready. Need to reconcile them", format.Pod(pod))
 		return true, sandboxStatus.Metadata.Attempt + 1, sandboxStatus.Id
 	}
-	if sandboxStatus.State != runtimeapi.PodSandboxState_SANDBOX_READY {
+
+	//Needs to create a new when not all container succeeded and the ready sandbox is not the latest one
+	if len(pod.Spec.Containers) != succeededContainerCount && sandboxStatus.State != runtimeapi.PodSandboxState_SANDBOX_READY {
```
Comment (tallclair) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go]: Just return early if `len(pod.Spec.Containers) == succeededContainerCount`

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go [hunk 5]
```diff
@@ -417,25 +417,39 @@ func (m *kubeGenericRuntimeManager) podSandboxChanged(pod *v1.Pod, podStatus *ku
...
+	succeededContainerCount := 0
+	for _, container := range pod.Spec.Containers {
+		for _, status := range podStatus.ContainerStatuses {
+			if container.Name == status.Name {
+				if status.State == kubecontainer.ContainerStateExited && status.ExitCode == 0 {
+					succeededContainerCount++
+				}
+			}
+		}
+	}
```
Comment (tallclair) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go]: Move this block after the readySandboxCount check

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go [hunk 6]
```diff
@@ -417,25 +417,39 @@ func (m *kubeGenericRuntimeManager) podSandboxChanged(pod *v1.Pod, podStatus *ku
 		}
 	}
 
-	// Needs to create a new sandbox when readySandboxCount > 1 or the ready sandbox is not the latest one.
+	//check the pod is succeeded
+	succeededContainerCount := 0
+	for _, container := range pod.Spec.Containers {
+		for _, status := range podStatus.ContainerStatuses {
+			if container.Name == status.Name {
+				if status.State == kubecontainer.ContainerStateExited && status.ExitCode == 0 {
```
Comment (tallclair) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go]: I think this depends on the restart policy? If the restart policy is `Always`, the sandbox should still be recreated. If the restart policy is `Never`, then failed containers should be included.

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go (hunk 5 above)
Comment (liucimin) [reply pkg/kubelet/kuberuntime/kuberuntime_manager.go]: Ok,I have move this block after the readySandboxCount check.

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go (hunk 6 above)
Comment (liucimin) [reply pkg/kubelet/kuberuntime/kuberuntime_manager.go]: Yes.
Now the code show us that:
when the restart policy is not  Always and the all container is succeded, the sandbox will not create a new one.
Others will enter the old process.

**Comment by liucimin** [comment]
friendly ping @tallclair

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go [hunk 7]
```diff
@@ -417,12 +417,27 @@ func (m *kubeGenericRuntimeManager) podSandboxChanged(pod *v1.Pod, podStatus *ku
...
 
-	// Needs to create a new sandbox when readySandboxCount > 1 or the ready sandbox is not the latest one.
+	// Needs to create a new sandbox when readySandboxCount > 1.
 	sandboxStatus := podStatus.SandboxStatuses[0]
 	if readySandboxCount > 1 {
 		klog.V(2).Infof("More than 1 sandboxes for pod %q are ready. Need to reconcile them", format.Pod(pod))
 		return true, sandboxStatus.Metadata.Attempt + 1, sandboxStatus.Id
 	}
+
+	//check the pod is succeeded
```
Comment (tallclair) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go]: nit: Space after `//` (everywhere)

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go [hunk 8]
```diff
@@ -417,12 +417,27 @@ func (m *kubeGenericRuntimeManager) podSandboxChanged(pod *v1.Pod, podStatus *ku
...
+	succeededContainerCount := 0
+	for _, container := range pod.Spec.Containers {
+		if containerSucceeded(&container, podStatus) {
```
Comment (tallclair) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go]: The logic for this method looks wrong to me... I think the container should only be considered succeeded if it's state is `ContainerStateExited`.

/cc @yujuhong

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go (hunk 7 above)
Comment (liucimin) [reply pkg/kubelet/kuberuntime/kuberuntime_manager.go]: Thx.

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go (hunk 8 above)
Comment (liucimin) [reply pkg/kubelet/kuberuntime/kuberuntime_manager.go]: Do you mean that there is a problem with the logic of the function of `containerSucceeded`?

Comment (tallclair) [reply pkg/kubelet/kuberuntime/kuberuntime_manager.go]: Yes. In the context we used it before, it didn't matter, but it would in this context.
